.genai/io/capstone/rag/embeddings.py

```
@@ -0,0 +1,212 @@
+"""
+Embeddings wrappers for generating vector representations of code.
+
+Supports:
+- OpenAI (text-embedding-3-small)
+- Gemini (gemini-embedding-001)
+"""
+
+from typing import List, Protocol
+from abc import ABC, abstractmethod
+
+
+class EmbeddingsBase(ABC):
+    """Abstract base class for embeddings providers."""
+
+    @abstractmethod
+    def count_tokens(self, text: str) -> int:
+        """Count the number of tokens in a text string."""
+        pass
+
+    @abstractmethod
+    def truncate_to_tokens(self, text: str, max_tokens: int = None) -> str:
+        """Truncate text to fit within token limit."""
+        pass
+
+    @abstractmethod
+    def embed_text(self, text: str) -> List[float]:
+        """Generate embedding for a single text."""
+        pass
+
+    @abstractmethod
+    def embed_batch(self, texts: List[str]) -> List[List[float]]:
+        """Generate embeddings for multiple texts efficiently."""
+        pass
+
+    @property
+    @abstractmethod
+    def dimension(self) -> int:
+        """Return the embedding dimension."""
+        pass
+
+
+class GeminiEmbeddings(EmbeddingsBase):
+    """Wrapper for Google's Gemini embedding API (free tier available)."""
+
+    MODEL = "models/gemini-embedding-001"
+    MAX_TOKENS = 2048  # Conservative limit for Gemini
+    DIMENSION = 768  # Using 768 for efficiency (supports 768, 1536, 3072)
+    BATCH_SIZE = 100  # Gemini batch limit
+
+    def __init__(self, api_key: str, dimension: int = 768):
+        """
+        Initialize the Gemini embeddings client.
+
+        Args:
+            api_key: Google AI API key (from ai.google.dev)
+            dimension: Output dimension (768, 1536, or 3072)
+        """
+        import google.generativeai as genai
+        genai.configure(api_key=api_key)
```

60 : The `__init__` method accepts a `dimension` parameter but doesn't validate it against the supported values mentioned in the docstring (768, 1536, 3072). An invalid value will cause a cryptic API error later on. Consider validating the input and raising a `ValueError` to fail fast if an unsupported dimension is provided.